transit_odp/browse/timetable_visualiser.py

```diff
@@ -0,0 +1,275 @@
+import logging
+
+from django.db.models import (
+    Q,
+    F,
+    CharField,
+)
+from django.db.models.functions import Coalesce
+
+from transit_odp.transmodel.models import (
+    Service,
+    ServicedOrganisationVehicleJourney,
+    OperatingDatesExceptions,
+    NonOperatingDatesExceptions,
+)
+from transit_odp.timetables.utils import (
+    get_vehicle_journeys_operating_nonoperating,
+    filter_df_serviced_org_operating,
+    get_df_operating_vehicle_journey,
+    get_df_timetable_visualiser,
+)
+import pandas as pd
+
+logger = logging.getLogger(__name__)
+logger.setLevel(logging.DEBUG)
+
+
+class TimetableVisualiser:
+
+    """
+    Timetable visualiser for the capturing of the time and the stops with journey codes
+    for the specific revision id, service code, line running on the target date on the
+    routes.
+    """
+
+    def __init__(self, revision_id, service_code, line_name, target_date) -> None:
+        """
+        Intializes the properties of the class.
+        """
+        self._revision_id = revision_id
+        self._service_code = service_code
+        self._line_name = line_name
+        self._target_date = target_date
+        self._day_of_week = target_date.strftime("%A")
+
+    def get_df_op_exceptions_vehicle_journey(self) -> pd.DataFrame:
+        """
+        Get the dataframe of vehicle journey which are operating on the target date
+        """
+        columns = ["vehicle_journey_id", "operating_date"]
+        qs_vehicle_journey_op_exceptions = (
+            OperatingDatesExceptions.objects.filter(operating_date=self._target_date)
+            .annotate()
+            .values(*columns)
+        )
+
+        return pd.DataFrame.from_records(qs_vehicle_journey_op_exceptions)
+
+    def get_df_nonop_exceptions_vehicle_journey(self) -> pd.DataFrame:
+        """
+        Get the dataframe of vehicle journey which are not operating on the target date
+
+        """
+        columns = ["vehicle_journey_id", "non_operating_date"]
+        qs_vehicle_journey_nonop_exceptions = (
+            NonOperatingDatesExceptions.objects.filter(
```

Comment:
Could we filter records based on vehicle journey id's obtained from the base query which will avoid scanning the entire table?